### PR TITLE
ICM (Initialisation, Control, Measurement) model for circuits

### PR DESCRIPTION
**Pull request by quantshah**
A representation of a quantum circuit consiting of qubit initialisation (I), controlled NOT gates (C) and measurements (M) with respect to different bases. According to [1], the transformation of a circuit to the ICM representation provides a cannonical form for an exact and fault-tolerant, error corrected circuit needed for optimization prior to final implementation on a hardware model.

The algorithm consists of magic qubit initializations, gate teleportation and conditional measurements on qubits. It can be a starting point for topological descriptions of quantum cirucuits important for fault-tolerant quantum computation and can become a part of a pipeline for optimization of topological quantum computing circuits. Check out the game (MeQuanics)[http://www.mequanics.com.au] which aims to do so using crowd sourcing. Several proposals exist for circuit optimizations once we have the topological circuit model (Machine learning, monte carlo search etc) and the ICM model can be a good starting point to get to the initial topological circuit (as far as I could understand). 

The algorithm can be roughly described as follows : 

```
Require: Circuit C composed from {Toffoli,CNOT,H,P,T}
Replace the Toffoli gates with their decomposition
Replace the H gates with P V P (already implemented using `resolve_gates`)
Implement P V P using gate teleportation outlined in [1]
```

The final output is a quantum circuit only using qubit initialisations, CNOT and conditional measurements/correction. Once that is complete, maybe another addition can be a conversion to the topological model and visualizations.

This is a work in progress.

TODO
------
[ ] Fix code errors
[ ] Write tests
[ ] Update examples and docs

References
----------
.. [1] arXiv:1509.03962v1 [quant-ph]

### CODE REVIEW

**Comment by sahmed95** [comment]
I had to move faster and complete the full implementation in a rather quick and unstructured manner for my lab. I will finish it and then systematically build this PR so that it becomes easy for review. I have a working implementation of the ICM model now and a partial implementation of [Pauli tracking](https://www.date-conference.com/files/proceedings/2014/pdffiles/05.6_7_ip2-19.pdf) which pushes all the corrections and measurement to the end of the circuit. I have written a quick example which demonstrates the use of this :

https://github.com/sahmed95/notebooks/blob/icm/qutip_icm/icm_pauli_tracked.ipynb

### Circuit with a rotation gate (T gate)

t_example

### Pauli tracked ICM representation of circuit

t_pauli_tracked

Thus, we get a quantum circuit with qubit initializations in the begining, a mesh of CNOT gates in the middle and measurements in the end. This seems to be very useful as the starting point of topological quantum circuit optimization.

Future work: 

1. Represent qubit initializations in some way. Right now I consider everything as a gate. (Perhaps make new objects for qubit intializations, measurement, correction etc.)

2. Better visualization such that all initializations, measurements and corrections can be stacked in one line.

3. Representing the information regarding P_dagger, T_dagger, V_dagger gate teleportation in some way. The difference is just based on the interpretation of the measurement.

4. Complete the Pauli tracking algorithm to get the gate sequence for correction based on measurement.

Tests, tests and more tests.

**Comment by ajgpitch** [comment]
This is all looking very nice! Looks like you have been making good
progress.

**Comment by sahmed95** [comment]
Hi, I have an implementation of the ICM model now and as a bonus, I got an introduction to topological quantum computing as well as a nice topological quantum computing circuit viewer.

I have it all in this repo -  https://github.com/sahmed95/icm

It needs major refactoring, testing and there is a lot of brute code in there so please ignore it for now. I had to finish this in the span of a few days to start work on something related to the same and hence the code isn't pretty. I will work on this after finishing up my winter internship and push the commits here.

An example can be found in https://github.com/sahmed95/icm/blob/master/tqc_icm/icm_example.ipynb

If this code has to go into qip, I guess I need someone with experience in this for review. How about https://github.com/alexandrupaler. The ICM paper is by him and he has implemented it in C++ in https://github.com/alexandrupaler/tqec.

**Comment by sahmed95** [comment]
Also just to throw in some nice images:

## Original circuit

original

## ICM conversion with gate teleportation

tqcc

## TQC using the TQC viewer

tqc

**Comment by ajgpitch** [comment]
On the face of it it all looks good to me. If you could find someone with expertise to review it, then that would be great.

**Comment by sahmed95** [comment]
Hi, this is ready for review. It probably needs a lot of refactoring but I have documented all functions and provided unit tests. I made a simple script for my lab where you can specify a circuit in a text file and the outputs will be generated. You can check it out here: https://github.com/sahmed95/icm

I have an example describing the use of this module. Where do the examples go ? @ajgpitch @nonhermitian 

As I mentioned, @alexandrupaler should be the right person for review. Dr. Devitt also mentioned Damian from ProjectQ (?) who could help perhaps.

**Comment by alexandrupaler** [comment]
Happy to review as soon as possible, but I cannot promise any deadline.

**Comment by sahmed95** [comment]
Thank you very much. I have documented most of the code and written simple unit tests. I would be happy to answer any questions regarding the code. If it helps, I can explain things over a short hangout session. Suggestions and comments are welcome.

**Comment by ajgpitch** [comment]
This has been awaiting a review for a long time. @sahmed95 can we chase this up?

**Comment by alexandrupaler** [comment]
@ajgpitch: i apologise. my mistake. i forgot about the review.
@sahmed95: downloaded your code. i get an error "ModuleNotFoundError: No module named 'qutip.qip.icm'. where do I find this module?

**Comment by sahmed95** [comment]
@alexandrupaler Thank you for looking into this. You will have to download this branch and install it in order to find qutip.qip.icm. I think the following set of steps should do it. These are from this [stack overflow question](https://stackoverflow.com/questions/1911109/how-to-clone-a-specific-git-branch)

1. Clone my version of QuTiP using 
```
git clone https://github.com/sahmed95/qutip.git
```

2. Checkout the icm branch in the cloned folder
```
git checkout icm
```
3. Install QuTiP from that branch by going to cloned QuTiP folder
```
python setup.py install
```
Alternate way to do this is just copy paste the `qutip/qip/icm.py`,  `qutip/qip/circuit.py` and `qutip/tests/test_qubitcircuit.py` files into your local installation of QuTiP. 

There are examples in [`qutip/tests/test_qubitcircuit.py`](https://github.com/qutip/qutip/pull/603/files#diff-b5635e2eb23e985de1a2be07e2dc0f7f) which may be useful to see how the code works.

@ajgpitch @nonhermitian Suggestions on the recommended way of doing this ? There is this tool called ["hub"](https://hub.github.com) which allows you to directly checkout pull requests using `hub checkout https://github.com/qutip/qutip/pull/603`

**Comment by sahmed95** [comment]
@alexandrupaler, any updates ? The main part of the code is in the function [`to_icm()`](https://github.com/qutip/qutip/pull/603/files#diff-9b2fa79d9553701cbcf7c56a0aa1c679R539)

First, the Toffoli and H gates are decomposed and then the P, V, T gates are replaced with the representation you have in the paper using gate teleportation.

File: qutip/qip/icm.py [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+    """
+    gate = cicm.gates[idx]
+    if icm_label(gate.name, gate.arg_label) == "P":
+        t = gate.targets[0]
+        # Move targets and controls of all gates before the P gate
+        cicm.gates[idx] = Gate("Y", targets=[t + 1],
+                               arg_label=r"ancilla")
+
+        for gate in cicm.gates:
+            if gate.arg_label in ["ancilla",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+        # Move targets and controls of all gates after the P gate
+
+        for gate in cicm.gates[idx+4:]:
+            if gate.arg_label in ["ancilla", "measurement",
+                                  "correction", "teleported"]:
+                continue
+
+            if gate.targets[0] == t:
+                gate.targets[0] += 1
+
```
Comment (herr-d) [inline qutip/qip/icm.py]: You missed moving the control of CNOTS here.
(see error 1)

File: qutip/qip/icm.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+                          arg_label=r"correction")
+
+        cicm.gates.insert(idx + 1, CNOT)
+        cicm.gates.insert(idx + 2, measurement)
+        cicm.gates.insert(idx + 3, correction)
+
+        # Move targets and controls of all gates after the P gate
+
+        for gate in cicm.gates[idx+4:]:
+            if gate.arg_label in ["ancilla", "measurement",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+        The quantum circuit with P gates implemeted using Gate teleportation.
+    """
+    gate = cicm.gates[idx]
+    if icm_label(gate.name, gate.arg_label) == "V":
+        t = gate.targets[0]
+        # Move targets and controls of all gates before the P gate
+        cicm.gates[idx] = Gate("Y", targets=[t + 1],
+                               arg_label=r"ancilla")
+        for gate in cicm.gates:
+            if gate.arg_label in ["ancilla",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        correction = Gate("x/z", controls=[t], targets=[t + 1],
+                          arg_label=r"correction")
+
+        cicm.gates.insert(idx + 1, CNOT)
+        cicm.gates.insert(idx + 2, measurement)
+        cicm.gates.insert(idx + 3, correction)
+
+        # Move targets and controls of all gates after the P gate
+        for gate in cicm.gates[idx+4:]:
+            if gate.arg_label in ["ancilla", "measurement",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+    -------
+    cicm: QubitCircuit
+        The quantum circuit with P gates implemeted using Gate teleportation.
+    """
+    gate = cicm.gates[idx]
+    if icm_label(gate.name, gate.arg_label) == "T":
+        t = gate.targets[0]
+        # Move targets and controls of all gates before the T gate
+        for gate in cicm.gates:
+            if gate.arg_label in ["ancilla", "measurement",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+                          controls=[t+1], arg_label="correction"))
+        cicm.gates.insert(idx + 14, Gate("x/z", targets=[t + 3],
+                          controls=[t+2], arg_label="correction"))
+        cicm.gates.insert(idx + 15, Gate("z/x", targets=[t + 4],
+                          controls=[t+3], arg_label="correction"))
+
+        # Move targets and controls of all gates after the T gate
+
+        for gate in cicm.gates[idx+16:]:
+            if gate.arg_label in ["ancilla", "measurement",
```
Comment (herr-d) [inline qutip/qip/icm.py]: remove this check.
relates to error 2

File: qutip/qip/icm.py [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+
+        # Move targets and controls of all gates after the P gate
+        for gate in cicm.gates[idx+4:]:
+            if gate.arg_label in ["ancilla", "measurement",
+                                  "correction", "teleported"]:
+                continue
+
+            if gate.targets[0] == t:
+                gate.targets[0] += 1
+
```
Comment (herr-d) [inline qutip/qip/icm.py]: You missed moving the control of CNOTS here.
(see error 1)

File: qutip/qip/icm.py [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+        # Move targets and controls of all gates after the T gate
+
+        for gate in cicm.gates[idx+16:]:
+            if gate.arg_label in ["ancilla", "measurement",
+                                  "correction", "teleported"]:
+                continue
+            if gate.targets[0] == t:
+                gate.targets[0] += 5
+
```
Comment (herr-d) [inline qutip/qip/icm.py]: You missed moving the control of CNOTS here.
(see error 1)

File: qutip/qip/icm.py [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+        cicm.N += 5
+    return (cicm)
+
+
+def visualise(icm_circuit):
```
Comment (herr-d) [inline qutip/qip/icm.py]: This function is not used anywhere.
Should this be included in the class? Is it needed?

File: qutip/qip/tests/test_icm.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+
+    assert_(icm_representation.gates[0].name == "IN")
+    assert_(icm_representation.gates[2].name == "CNOT")
+    assert_(icm_representation.gates[2].controls[0] == 0)
+    assert_(icm_representation.gates[2].targets[0] == 6)
+
+    assert_(icm_representation.gates[-1].name == "OUT")
+    assert_(icm_representation.gates[-2].name == "OUT")
+
+    assert_(icm_representation.gates[-3].controls[0] == 0)
```
Comment (herr-d) [inline qutip/qip/tests/test_icm.py]: The control should be at qubit 5 and not 0.

relates to Error 1

**Comment by ajgpitch** [comment]
@sahmed95 How do you think this is shaping up? Getting near to merge-ready?

**Comment by ajgpitch** [comment]
Is this ready to merge now? I think it does not need to be perfect.

**Comment by quantshah** [comment]
@ajgpitch I looked into the changes that Daniel mentioned. But, I would like to close this without merging with QuTiP as I feel that the QubitCircuit class needs some new features to make this addition interesting (Eg., allowing special ancillas, controlled measurement and tracking of the Pauli measurement errors etc.)

This was almost 3 years back where I brute forced all of that as Gates. I am not happy to release this code as it is which can be buggy. So, I am closing this. If there is more interest later, we could open it again.

**Pull request closed without merging**